lib/rules/no-import-compiler-macros.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * @author Wayne Zhang
+ * See LICENSE file in root directory for full license.
+ */
+'use strict'
+
+const COMPILER_MACROS = new Set([
+  'defineProps',
+  'defineEmits',
+  'defineExpose',
+  'withDefaults'
+])
+
+const VUE_MODULES = new Set(['@vue/runtime-core', '@vue/runtime-dom', 'vue'])
+
+/**
+ * @param {Token} node
+ */
+function isComma(node) {
+  return node.type === 'Punctuator' && node.value === ','
+}
+
+/**
+ * @param {Token} node
+ */
+function isLeftCurlyBrace(node) {
+  return node.type === 'Punctuator' && node.value === '{'
+}
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'disallow importing vue compiler macros',
+      categories: undefined,
+      url: 'https://eslint.vuejs.org/rules/no-import-compiler-macros.html'
+    },
+    fixable: 'code',
+    schema: [],
+    messages: {
+      noImportCompilerMacros:
+        "'{{name}}' is a compiler macro and no longer needs to be imported from '{{source}}'."
+    }
+  },
+  /**
+   * @param {RuleContext} context
+   * @returns {RuleListener}
+   */
+  create(context) {
+    const sourceCode = context.getSourceCode()
+
+    return {
+      ImportDeclaration(node) {
+        if (node.specifiers.length === 0) return
+
+        if (VUE_MODULES.has(node.source.value)) {
```

Comment:
Better return early here, too, to remove one level of indentation and make the rest of the block easier to read:

```ts
        if (node.specifiers.length === 0 || !VUE_MODULES.has(node.source.value)) {
          return
        }
```